applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/eigenfrequency_response_function_utility.h

```diff
@@ -64,327 +64,434 @@ namespace Kratos
+                for(std::size_t i = 0; i < num_of_traced_eigenfrequencies; i++)
+                {
+                    aux_matrix = perturbed_LHS;
+                    aux_matrix -= (perturbed_mass_matrix * traced_eigenvalues[i]);
+                    aux_vector = prod(aux_matrix , eigenvectors_of_element[i]);
+                    gradient_contribution[2] += gradient_prefactors[i] * inner_prod(eigenvectors_of_element[i] , aux_vector) * mWeightingFactors[i];
+
+                    aux_matrix = Matrix(0,0);
+                    aux_vector = Vector(0);
+                }
+
+                node_i.Z0() -= mDelta;
+                // End derivative of response w.r.t. z-coord --------------------
 
-				// Derivative of response w.r.t. x-coord ------------------------
-				node_i.X0() += mDelta;
+                // Assemble sensitivity to node
+                noalias(node_i.FastGetSolutionStepValue(SHAPE_SENSITIVITY)) += gradient_contribution;
 
-				elem_i.CalculateMassMatrix(perturbed_mass_matrix, CurrentProcessInfo);
-				perturbed_mass_matrix = (perturbed_mass_matrix - mass_matrix_org) / mDelta;
-				perturbed_mass_matrix *= eigenvalue;
+            }// End loop over nodes of element
 
-				elem_i.CalculateLocalSystem(perturbed_LHS, dummy ,CurrentProcessInfo);
-				perturbed_LHS = (perturbed_LHS - LHS_org) / mDelta;
+        }// End loop over elements
+    }
 
-				perturbed_LHS -= perturbed_mass_matrix;
+    // --------------------------------------------------------------------------
+    double GetEigenvalue(int eigenfrequency_id)
+    {
+        const int num_of_computed_eigenvalues = (mrModelPart.GetProcessInfo()[EIGENVALUE_VECTOR]).size();
+        return (mrModelPart.GetProcessInfo()[EIGENVALUE_VECTOR])[eigenfrequency_id-1];
+    }
 
-				aux = prod(perturbed_LHS , eigenvector_of_element);
-				gradient_contribution[0] = inner_prod(eigenvector_of_element , aux);
+    // --------------------------------------------------------------------------
+    Vector GetEigenvectorOfElement(ModelPart::ElementType& traced_element, int eigenfrequency_id, int size_of_eigenvector)
+    {
+        Vector eigenvector_of_element(size_of_eigenvector,0.0);
 
-				node_i.X0() -= mDelta;
-				// End derivative of response w.r.t. x-coord --------------------
+        // Get eigenvector of element
+        int k = 0;
+        const int NumNodeDofs = size_of_eigenvector/traced_element.GetGeometry().size();
+        for (auto& node_i : traced_element.GetGeometry())
+        {
+            Matrix& rNodeEigenvectors = node_i.GetValue(EIGENVECTOR_MATRIX);
 
-				// Reset pertubed RHS and mass matrix
-				perturbed_LHS = Matrix(0,0);
-				perturbed_mass_matrix = Matrix(0,0);
-				aux = Vector(0);
+            for (int i = 0; i < NumNodeDofs; i++)
+                eigenvector_of_element(i+NumNodeDofs*k) = rNodeEigenvectors((eigenfrequency_id-1),i);
 
-				// Derivative of response w.r.t. y-coord ------------------------
-				node_i.Y0() += mDelta;
+            k++;
```

Comment:
Just my personal opinion: I always try to avoid counters, because one easily forgets to increment them.

one could use:
- ```&node_i  - &traced_element.GetGeometry()[0]``` instead of ```k```
or 
- a standard for loop over the ```traced_element.GetGeometry().size()``` 

I don`t know what is more efficient though.